agent/ecscni/namespace_helper_windows.go

```diff
@@ -0,0 +1,199 @@
+	if config.BlockInstanceMetadata {
+		if taskENI == nil {
+			return errors.New("failed to configure firewall due to invalid task eni")
+		}
+
+		firewallRuleName := fmt.Sprintf(blockIMDSFirewallRuleNameFormat, taskENI.GetPrimaryIPv4Address())
+
+		// Delete the firewall rule created for blocking IMDS access by the task.
+		checkExistingFirewallRule := fmt.Sprintf(checkExistingFirewallRuleCmdFormat, firewallRuleName)
+		blockIMDSFirewallRuleDeletionCmd := fmt.Sprintf(deleteFirewallRuleCmdFormat, firewallRuleName)
+
+		// The separator would be "&&" to ensure if the firewall rule exists then delete it.
+		// An error at this point means that the firewall rule was not present and was therefore not deleted.
+		// Hence, skip returning the error as it is redundant.
+		nsHelper.invokeCommandsOnHost(execCmdsOnHostFn, []string{checkExistingFirewallRule, blockIMDSFirewallRuleDeletionCmd}, " && ")
+	}
+
+	return nil
+}
+
+// invokeCommandsInsideContainer executes a set of commands inside the container namespace.
+func (nsHelper *helper) invokeCommandsInsideContainer(ctx context.Context, containerID string, commands []string, separator string) error {
+
+	seelog.Debugf("[ECSCNI] Executing commands inside container %s namespace: %v", containerID, commands)
+
+	// Concatenate all the commands into a single command.
+	execCommands := strings.Join(commands, separator)
+	// Prepare the config command.
+	cfgCommand := []string{"cmd", "/C", execCommands}
+
+	execCfg := types.ExecConfig{
+		Detach: false,
+		Cmd:    cfgCommand,
+		User:   containerAdminUser,
+	}
+
+	execRes, err := nsHelper.dockerClient.CreateContainerExec(ctx, containerID, execCfg, dockerclient.ContainerExecCreateTimeout)
+	if err != nil {
+		seelog.Errorf("[ECSCNI] Failed to execute command in container %s namespace [create]: %v", containerID, err)
+		return err
+	}
+
+	err = nsHelper.dockerClient.StartContainerExec(ctx, execRes.ID, types.ExecStartCheck{Detach: false, Tty: false},
+		dockerclient.ContainerExecStartTimeout)
+	if err != nil {
+		seelog.Errorf("[ECSCNI] Failed to execute command in container %s namespace [pre-start]: %v", containerID, err)
+		return err
+	}
+
+	// Query the exec container to determine if the commands succeeded.
+	inspect, err := nsHelper.dockerClient.InspectContainerExec(ctx, execRes.ID, dockerclient.ContainerExecInspectTimeout)
+	if err != nil {
+		seelog.Errorf("[ECSCNI] Failed to execute command in container %s namespace [inspect]: %v", containerID, err)
+		return err
+	}
+	// If the commands fail then return error.
+	if !inspect.Running && inspect.ExitCode != 0 {
+		return errors.Errorf("commands failed inside container %s namespace: %d", containerID, inspect.ExitCode)
+	}
+
+	return nil
+}
+
+// invokeCommandsOnHost invokes given commands on the host instance using the executeFn.
+func (nsHelper *helper) invokeCommandsOnHost(executeFn execCmdExecutorFn, commands []string, separator string) error {
```

Comment:
why not store executeFn as a field of nsHelper? i think that way it's easier to mock it in test and you don't need to have executeFn as an argument like this

Reply:
Yes that is functionally equivalent. Considering that if Linux implementation of the same is required sometime, we will have similar function signature, we can move it inside helper struct.
I'll make these changes.